biorxiv/journal_tracker/journal_recommendation_experiment.py

```diff
@@ -0,0 +1,354 @@
+        .drop("pmcid", axis=1)
+        .groupby("journal", group_keys=False)
+        .apply(lambda x: x.sample(min(len(x), 100), random_state=100))
+        .set_index("document")
+    )
+    for dim in pmc_embedding_dict
+}
+
+
+# # Train Recommendation System
+
+# In[11]:
+
+
+knn_model = KNeighborsClassifier(n_neighbors=10)
+
+
+# ## Random Journal Prediction
+
+# The central idea here is to answer the question what is the accuracy when journals are recommended at random?
+
+# In[12]:
+
+
+model = DummyClassifier(strategy='uniform')
+
+
+# In[14]:
+
+
+_ = cross_validation(
+    model, subsampled_training_dataset[300], 
+    dummy_evaluate, cv=10, 
+    random_state=100, top_predictions=10
+)
+
+
+# ## KNearestNeighbors Paper by Paper Comparison - Full Dataset
+
+# Assuming I didn't take mega-journal influence into account, what would the initial recommendation accuracy be?
+
+# In[ ]:
+
+
+_ = cross_validation(
+    knn_model, full_training_dataset[300], 
+    knn_evaluate, cv=10,
+    random_state=100
+)
+
+
+# ## KNearestNeighbors Paper by Paper Comparison Subsampled
+
+# The first idea for a classifier is to compare which papers are similar to other papers. Due to the overflow of PLOS One papers I sub-sampled each journal to have only 100 papers for representation. Then trained a KNearestNeighbors to determine how often does the correct journal appear in the top ten neighbors as well as top twenty neighbors.
+
+# In[17]:
+
+
+result_dict = {}
+for dim in subsampled_training_dataset:
+    print(dim)
+    
+    fold_predictions = cross_validation(
+        knn_model, subsampled_training_dataset[dim], 
+        knn_evaluate, cv=10, 
+        random_state=100
+    )
+    
+    print()
+    
+    result_dict[dim] = fold_predictions
+
+
+# ## KNearestNeighbors Centroid analysis
+
+# Following up on the original idea, I thought a helpful experiment would be to perform a centroid analysis (i.e. take the average of all papers within each journal). Similar to above I trained a KNearestNeighbors classifier to see if the correct journal will appear in the top 10 neighbors.
+
+# In[16]:
+
+
+_ = cross_validation(
+    knn_model, subsampled_training_dataset[300], 
+    knn_centroid_evaluate, cv=10, 
+    random_state=100
+)
+
+
+# ## KNearestNeighbors Centroid Analysis - Full dataset
+
+# This section I'm using the entire dataset to calculate journal centroids and then evaluate performance on the sub-sampled dataset.
+
+# In[19]:
+
+
+predictions, true_labels = (
+    knn_centroid_evaluate(
+        knn_model,
+        full_training_dataset[300], 
+        subsampled_training_dataset[300]
+    )
+)
+
+
+# In[20]:
+
+
+accs = [
+    (
+         1 if true_labels[data_idx] in prediction_row 
+         else 0 
+     )
+     for data_idx, prediction_row in enumerate(predictions)
+]
+
+
+# In[21]:
+
+
+print(f"{np.sum(accs)} out of {len(accs)}")
+print(f"{np.mean(accs)}% correct")
```

Comment:
I'm outside of my field on this, do you think 40% accuracy pretty good in this case, given the 9K journals you're using

Reply:
> I'm outside of my field on this, do you think 40% accuracy pretty good in this case, given the 9K journals you're using

Well it is subsampled to 2K as I cut off journals with low representation (<100). Since this is the first time this analysis has been done, 40% is acceptable. I'm sure as soon as I release this to the public 40% will shoot up to possible >60% in a few years. Just a hard problem as paper content alone doesn't provide a robust way to predict journal endpoints. Predicting peer review is a challenge within itself.